### DISCUSSION TITLE
adding softreboot via crtl d

### DISCUSSION THREAD

**Discussion by OutOfTheBots** [pybricks/support#602] [Ideas]
I have used main line micropython for some time and a number of different port. Usually crtl d in the repl gives a soft reboot which is handy to clear all RAM and reset.

Can this be added to pybrick even it is only for spike and robot inventor that have enough space to add features

**Comment by dlech** [comment]
A soft reboot is performed every time the user program is stopped. And CTRL-D should already work in the REPL. What behavior are you seeing?

**Reply by OutOfTheBots** [reply]
see this short video to see behavior https://www.youtube.com/watch?v=lfT-hgrbahA

**Reply by laurensvalk** [reply]
https://github.com/pybricks/support/issues/232#issuecomment-1013862277 Is a bug, but the REPL is currently behaving as expected.

CTRL+D ends the REPL program. Just like any other program, it doesn’t restart automatically.

But if it isn’t too complicated to implement, it is perhaps indeed nice to restart the REPL automatically in this case.

I don’t use this often, so I don’t have a strong opinion on this one.

**Reply by dlech** [reply]
This isn't a documented feature (it may be removed or changed in the future), but you can currently press space 4 times to start the REPL again after CTRL+D if you don't want to click the button with the mouse.

**Reply by OutOfTheBots** [reply]
I am very new to pybricks but experienced with micropython. 

My suggestion would be for pybricks behavior should be as similar as possible to all the other ports of micropython.  

There is also the bigger issue of why the CRTL-D didn't reset the IUM that the only way that I could reset the IUM was turn off the hub and turn it back on again. In the history of mainline micropython there has been lots of bugs over time, usually if you hit a bug u can CRTL-D to reset and start again.

**Reply by laurensvalk** [reply]
It is already doing a soft-reset, so that wouldn't fix it. We prefer to just fix bugs instead of relying on reboots, though.

> There is also the bigger issue of why the CRTL-D didn't reset the IUM

During a soft-reset, hardware drivers are not reset as this would mean losing the bluetooth connection and losing sync with all motors and sensors. For the IMU, it would mean losing the yaw information (even though we don't have this yet). This behavior is consistent with the LEGO firmware by the way.

> My suggestion would be for pybricks behavior should be as similar as possible to all the other ports of micropython.

If you prefer a pure STM32 build with no drivers for LEGO motors and sensors, you could build the upstream MicroPython source code. For most users, this would be a lot harder to use, so our approach is a bit different.

**Reply by OutOfTheBots** [reply]
@laurensvalk 
Thanks for your reply.

A pure STM32 build **without any Lego drivers for motors or sensors** would be not very useful. I was suggesting as similar behavior because the skills that kids are learning on Lego are then more portable to other platforms. Micropython has become very popular, even Ardunio has paid developers from OpenMV to create a  micropython port for their new boards based upon the NRF52840 chip

**Reply by laurensvalk** [reply]
I suppose we'd need Arduino to sponsor Pybricks then :smile:

**Reply by OutOfTheBots** [reply]
> During a soft-reset, hardware drivers are not reset as this would mean losing the bluetooth connection and losing sync with all motors and sensors. For the IMU, it would mean losing the yaw information (even though we don't have this yet). This behavior is consistent with the LEGO firmware by the way.

I understand that during a soft reboot that you don't want to loose Bluetooth or USB connection to PC but in most cases the reason your doing a soft reboot to reset all motors, sensors, IUM data and clear all RAM.

Imagine running your program then doing a soft reboot then running your program a second time but now your robot's encoders and IUM isn't zeroed at the start up of your program but rather remembers data from before the soft reboot.

I would suggest that a soft reboot resets everything except connection to PC as it is  a pain the only way to reset everything is a complete hard reset

**Reply by dlech** [reply]
This pretty much happens already. IMU support is unfinished and really shouldn't be advertised as working yet so bugs are expected there. Motors are stopped/reset each time the user program stops (including CTRL+D in REPL). The MicroPython runtime is completely restarted, so all of the RAM used in the user program or in the REPL is freed.

**Reply by laurensvalk** [reply]
This is actually being done for the most part. In fact, much more so than what would happen in a standard micropython board.

For example, unlike the REPL in the LEGO firmware or some MicroPython boards, we take care of stopping the motor and getting them back in a clean state.

Basically, a hard reset should never be needed. If you find there’s a need, that indicates there’s a bug (as you’ve seen with the IMU, and it is logged as such).

As for the IMU, that’s actually one of the things that should not be reset. You wouldn’t want it to start calibrating just as your code begins, because this is typically when the robot is moving.

It could still make sense to have the yaw value start at 0 (this is what the heading method will do) but that has nothing to do with resetting the hardware or driver.

**Reply by OutOfTheBots** [reply]
@dlech  thanks for the clarification on the soft reboot.

> As for the IMU, that’s actually one of the things that should not be reset. You wouldn’t want it to start calibrating just as your code begins, because this is typically when the robot is moving.

I think this is more fuel for the idea that the IMU shouldn't start on boot up but rather the user has to call an initialization method something like IMU.init(orientation=axis.X) that way the user can ensure the robot is still during the startup of the IUM and also specify which rotation axis needs to be tracked for yaw as the other 2 axis orientation is known by sensing the linear pull of gravity.

I am of the opinion that soft-reboot should also reset the IMU the same as the motors and sensors.
